### DISCUSSION TITLE
Monitoring Memory Per VipsImage

### DISCUSSION THREAD

**Discussion by joshuamsager** [libvips/libvips#2442] [Help wanted]
Our app uses the following pattern to monitor how long a transformation takes to occur using the `eval` signal.

Something like:

```c
void timeout_callback(VipsImage *image, gpointer progress, gpointer data) 
{
    time_t now = time(NULL);
    time_t *expiration = (time_t *) data;

    if (now > *expiration) {
        vips_image_set_kill(image, true);
    }
}

int with_timeout(VipsImage *image, int timeout) 
{
    time_t now = time(NULL);
    time_t expiration = now + timeout;

    vips_image_set_progress(image, true);
    return g_signal_connect(image, "eval", G_CALLBACK( timeout_callback ), &expiration);
}

```

Is there something you can suggest we implement in order to monitor how much memory a transformation uses?

This would be useful in cloud environments where our application is terminated if it exceeds memory limits. The idea here is that we could limit transformations to some arbitrary size so that resources are not abused. Without this, a transform taking more than 10 GB (seen from experience) gets OOMKilled and can become a DOS attack vector.

**Comment by kleisauke** [comment]
Oh, funny. I implemented something similar this evening as well. See for example commit https://github.com/weserv/images/commit/53365afa9ac135f7a5a9365099b3b1b3eed151b8.

As for monitoring memory usage, I think you could use [`vips_tracked_get_mem()`](https://www.libvips.org/API/current/libvips-memory.html#vips-tracked-get-mem), but this would not include allocations for external libraries.

**Reply by joshuamsager** [reply]
`vips_tracked_get_mem()` has been really helpful for tracking overall memory usage, but doesn't help for tracking specific VipsImages, unfortunately.

I was hoping VipsImage would expose some sort of field like `memory_footprint` which we could then check if it exceeds some limit (maybe 1 GB) in a callback.

**Comment by jcupitt** [comment]
Hello @joshuamsager,

Opening an image is always fast, so applications can check the image dimensions before processing.

It should always be enough to spot memory bombs -- I think we'd consider it a bug if you find a case where that doesn't work.

**Reply by joshuamsager** [reply]
In general, I agree. There are outliers, however.

For instance, this < 4 MB gif takes 1.63 GB to copy:
```bash
joshuasager@libvips (master) $ vips copy sample.gif[n=-1] out.gif
memory: high-water mark 1.63 GB
joshuasager@libvips (master) $ ls -l sample.gif
-rw-r--r--  1 joshuasager  staff  3449741 Sep  4 14:57 sample.gif
```

I am all for fixing these edge cases in libvips 100% but would also love some mechanism to defend against these one-offs in a production environment. It also makes our integration with libvips more resilient as libvips evolves over time.

**Reply by jcupitt** [reply]
GIF save is a bit of a special case --- vips 8.11 delegates to imagemagick for this, so memory use is high, speed is not great, and libvips has little control over the process.

Lovell has made a new libvips GIF save module for 8.12 and it's at least 10x faster and uses 10x less memory, so hopefully this specific problem should go away. 

I see:

```
$ /usr/bin/time -f %M:%e vips copy x.gif[n=-1] y.gif
253920:10.16
$ ls -l x.gif
-rw-rw-r-- 1 john john 30948111 May 19 14:39 x.gif
```

So a 30mb GIF copies in 10s and needs 250mb of memory. It's in libvips GIT master if you'd like to try it.

**Reply by lovell** [reply]
I recognise that `sample.gif` file. It's a 301 frame image, with each frame being 1080x1080. Part of the problem with saving GIFs is that the whole image has to be in memory to generate the palette. I did start work on frame-local palettes - see commit https://github.com/lovell/libvips/commit/2e5e2f4a8fc37b1049748ab32877eba0e37fd498 - but these are slower to generate and create larger files, however peak memory usage is much lower as only one frame at a time needs to be in memory.

For large images with hundreds of frames such as this example, if we were to optionally provide frame-local palettes *and* reduce the bitdepth from 8 to perhaps 6, so using 64 colours per frame palette, then I think that might be enough to offset the otherwise increased quantisation time and file size. (I'll need to update that WIP commit as it won't apply cleanly to the master branch.)

**Reply by jcupitt** [reply]
Ouch, that does sound like a pathological case. My 30mb GIF is 176x112 and 1,900 frames, and that seems to perform OK, fwiw.

**Reply by joshuamsager** [reply]
I wonder if there is some simple math we can use to predict how much memory libvips needs to process the entire file. This might be enough for our use case. If we can estimate how much memory a transform might take, we can make decisions on whether or not to proceed.

In this scenario, 301 frames x 1080 width x 1080 height = 351,086,400 N_PELS or 351 MB -- Still shy from 1.4 GB.

**Reply by jcupitt** [reply]
libvips itself scales very approximately like this:

```
CPU = k * image.width * image.height / n_threads + c
mem = k * image.width * n_threads + c
```

ie. memory scales with image width, CPU with pixels. 

Of course many format libraries have other scaling properties which we have little control over. As Lovell said. GIF save needs the whole image uncompressed in memory so it can compute a global palette. Your RGB image will need `image.width * image.height * image.bands` bytes, or about 1gb at minimum.

**Reply by lovell** [reply]
I've updated the WIP branch for frame-local GIF palettes via commit https://github.com/lovell/libvips/commit/c598208226f1ff64ff7ac404c0a4f6399849e2e1 so it will apply cleanly to the master branch. Sadly it still takes ~2x longer to quantise than using a global palette, even when reducing the bitdepth and dithering, so I'm unsure the extra complexity is worth it.

**Reply by jcupitt** [reply]
Oh, that's a shame :( 

I suppose we'd need to dig into libimagequant for further improvements.

How about adding an option to generate the colormap from just the first N frames, with N == -1 meaning the whole image (current behaviour)? 

We'd need to be able to call the dither and the palette parts of libimagequant separately. I've no idea if this is possible!

I suppose you could compute a new palette every N frames as well, maybe that would be better?

Or better still, perhaps automatically compute a new palette when the new frame is very different from the old frame.

**Reply by jcupitt** [reply]
It looks like libimagequant does allow separate palette gen and dither. Shall I have a go at a patch that only recomputes the palette occasionally?

**Reply by joshuamsager** [reply]
FYI: I have tested https://github.com/libvips/libvips/pull/2445 locally and it seems to be working well. The speed tradeoff is a shame, however, this works best in a cloud environment where OOMKills become a problem if transformations take too much memory.
